### PR TITLE
Fix files with comments only

### LINKED ISSUE #806: Blank lines are lost in files that contain only comments
Input code:

```javascript
/**
 * @typedef {DataDrivenMapping|ConstantMapping} Mapping
 */
/**
 * @typedef {Object.<String, Mapping>} ConfigurationMapping
 */

/**
 * @typedef {Function} D3Scale - a D3 scale
 * @property {Function} ticks
 * @property {Function} tickFormat
 */
```

The blank line in the middle, which separates groups of related [JSDoc `@typedef`s](http://usejsdoc.org/tags-typedef.html), is lost [when Prettier reformats][output]:

```javascript
/**
 * @typedef {DataDrivenMapping|ConstantMapping} Mapping
 */
/**
 * @typedef {Object.<String, Mapping>} ConfigurationMapping
 */
/**
 * @typedef {Function} D3Scale - a D3 scale
 * @property {Function} ticks
 * @property {Function} tickFormat
 */
```

I would expect the blank line to be preserved, the same way Prettier preserves single blank lines between blocks of code. These JSDoc `/** */` comments basically *are* blocks of code that just happen to syntactically be comments.

### A more minimal example

Input:

```javascript
// a

// b
```

The blank line [is still lost][minimal-output]:

```javascript
// a
// b
```

But adding non-comment code anywhere in the file causes the blank line to be preserved:

```javascript
code();
// a

// b
```

The above [stays unchanged][minimal-working-output] after reformatting.

[output]: https://prettier.github.io/prettier/#%7B%22content%22%3A%22%2F**%5Cn%20*%20%40typedef%20%7BDataDrivenMapping%7CConstantMapping%7D%20Mapping%5Cn%20*%2F%5Cn%2F**%5Cn%20*%20%40typedef%20%7BObject.%3CString%2C%20Mapping%3E%7D%20ConfigurationMapping%5Cn%20*%2F%5Cn%5Cn%2F**%5Cn%20*%20%40typedef%20%7BFunction%7D%20D3Scale%20-%20a%20D3%20scale%5Cn%20*%20%40property%20%7BFunction%7D%20ticks%5Cn%20*%20%40property%20%7BFunction%7D%20tickFormat%5Cn%20*%2F%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22doc%22%3Afalse%7D%7D

[minimal-output]: https://prettier.github.io/prettier/#%7B%22content%22%3A%22%2F%2F%20a%5Cn%5Cn%2F%2F%20b%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22doc%22%3Afalse%7D%7D

[minimal-working-output]: https://prettier.github.io/prettier/#%7B%22content%22%3A%22code()%3B%5Cn%2F%2F%20a%5Cn%5Cn%2F%2F%20b%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22doc%22%3Afalse%7D%7D

### PR DESCRIPTION
**Pull request by yamafaktory**
Trying to fix #806.

### CODE REVIEW

**Comment by yamafaktory** [comment]
Not quite sure why the tests are adding an empty space...

**Comment by yamafaktory** [comment]
I tried with a different heuristic. This is definitely better as it maintains in-between empty lines in files containing only comments. But this still needs some refinements.

**Comment by yamafaktory** [comment]
Ok, one last thing I need to fix is when `use strict` is part of the AST.

**Comment by yamafaktory** [comment]
@vjeux / @jlongster can you please review it? This should be better now 🤞.

**Comment by yamafaktory** [comment]
Getting `There was an error while trying to fetch the log.` but the tests are passing for me locally (node v4&6).

**Comment by yamafaktory** [comment]
🎉

File: tests/flow/declare_export/__snapshots__/jsfmt.spec.js.snap [hunk 1]
```diff
@@ -586,6 +586,7 @@ export default {
  * @providesModule ES6Default
  * @flow
  */
+
```
Comment (yamafaktory) [inline tests/flow/declare_export/__snapshots__/jsfmt.spec.js.snap]: Now, this is fixed, the empty line is preserved.

File: tests/flow/es6modules/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -686,6 +686,7 @@ export default {
  * @providesModule ES6Default
  * @flow
  */
+
```
Comment (yamafaktory) [inline tests/flow/es6modules/__snapshots__/jsfmt.spec.js.snap]: Idem.

**Comment by yamafaktory** [comment]
@vjeux Do you have time to take a look at it? Thanks!

File: src/comments.js [hunk 3]
```diff
@@ -147,8 +149,10 @@ function attach(comments, ast, text, options) {
         handleMemberExpressionComments(enclosingNode, followingNode, comment) ||
         handleIfStatementComments(enclosingNode, followingNode, comment) ||
         handleTryStatementComments(enclosingNode, followingNode, comment) ||
+        handleClassComments(enclosingNode, comment) ||
```
Comment (Royce) [inline src/comments.js]: This looks like a duplicate line @yamafaktory

Comment (yamafaktory) [inline src/comments.js]: Yeah, I should fix it, this is coming from a merge conflict.

**Comment by yamafaktory** [comment]
Now should be clean and up-to-date again. @vjeux / @jlongster do you have time to review it before it becomes out of sync again 😄? Thanks!

File: src/comments.js [hunk 4]
```diff
@@ -152,14 +154,15 @@ function attach(comments, ast, text, options) {
         handleMemberExpressionComments(enclosingNode, followingNode, comment) ||
         handleIfStatementComments(enclosingNode, followingNode, comment) ||
         handleTryStatementComments(enclosingNode, followingNode, comment) ||
-        handleImportSpecifierComments(enclosingNode, comment) ||
         handleClassComments(enclosingNode, comment) ||
+        handleImportSpecifierComments(enclosingNode, comment) ||
```
Comment (vjeux) [inline src/comments.js]: can you put it back to the place it was before?

File: src/comments.js [hunk 5]
```diff
@@ -184,10 +187,11 @@ function attach(comments, ast, text, options) {
         ) ||
         handleImportSpecifierComments(enclosingNode, comment) ||
         handleTemplateLiteralComments(enclosingNode, comment) ||
-        handleCallExpressionComments(precedingNode, enclosingNode, comment) ||
         handleClassComments(enclosingNode, comment) ||
+        handleCallExpressionComments(precedingNode, enclosingNode, comment) ||
```
Comment (vjeux) [inline src/comments.js]: same

**Comment by vjeux** [comment]
lgtm! I'll merge it with the small movements :) Sorry for the very slow review.

**Pull request merged**